src/Workspaces/Core/Portable/CodeActions/CodeAction.cs

```diff
@@ -249,6 +249,17 @@ protected virtual Task<Document> GetChangedDocumentAsync(CancellationToken cance
             return document;
         }
 
+        /// <summary>
+        /// Called by the CodeActions on the UI thread to determine if the CodeAction is still 
+        /// applicable and should be presented to the user.  CodeActions can override this if they 
+        /// need to do any final checking that must be performed on the UI thread (for example
+        /// accessing and querying the Visual Studio DTE).
+        /// </summary>
+        protected internal virtual bool IsApplicable(Workspace workspace)
```

Comment:
ya, I dont like 1) option as well. but I think we have enough feedbacks asking to have a way to access UI thread in some way. so I think we should provide a way to user.